docs/root/version_history/current.rst

```diff
@@ -25,6 +25,7 @@ Bug Fixes
 * dns: fix a bug where custom resolvers provided in configuration were not preserved after network issues.
 * http: fixed URL parsing for HTTP/1.1 fully qualified URLs and connect requests containing IPv6 addresses.
 * http: sending CONNECT_ERROR for HTTP/2 where appropriate during CONNECT requests.
+* tls: fix detection of the upstream connection close event.
```

Comment:
Is this about the case where the proxy gets a graceful SSL shutdown via close notify, or the case where the peer closes the connection without sending close notify first?

Reply:
Correct.

Reply:
Would be good to cover both cases in tests.  Not sure how you want to handle the case without close notify.  Fairly sure there are some implementations out there that forget to send close notify.

Reply:
That's not an issue, I can trigger shutdown without close_notify in tests using `SSL_set_quiet_shutdown(..., 1)`.